Include/Rover/TrialView.hpp

```diff
@@ -0,0 +1,270 @@
+  */
+  template<typename S>
+  class TrialView {
+    public:
+
+      //! The type of the samples stored in the Trial.
+      using Sample = S;
+
+      //! Random access constant iterator.
+      class ConstIterator {
+        public:
+
+          //! Creates a constant iterator.
+          /*
+            \param t Trial to read.
+            \param offset Offset from t.begin().
+          */
+          template<typename Trial>
+          ConstIterator(const Trial& t, std::ptrdiff_t offset);
+
+          //! Increments the iterator.
+          ConstIterator& operator ++();
+
+          //! Increments the iterator.
+          ConstIterator operator ++(int);
+
+          //! Decrements the iterator.
+          ConstIterator& operator --();
+
+          //! Decrements the iterator.
+          ConstIterator operator --(int);
+
+          //! Advances the iterator.
+          ConstIterator& operator +=(std::ptrdiff_t offset);
+
+          //! Moves the iterator back.
+          ConstIterator& operator -=(std::ptrdiff_t offset);
+
+          //! Returns an iterator that is the result of advancing this.
+          ConstIterator operator +(std::ptrdiff_t offset);
+
+          //! Returns an iterator that is the result of moving this back.
+          ConstIterator operator -(std::ptrdiff_t offset);
+
+          //! Returns the offset between two iterators.
+          std::ptrdiff_t operator -(ConstIterator other) const;
+
+          //! Checks whether two iterators point to the same Sample.
+          bool operator ==(ConstIterator other) const;
+
+          //! Checks whether two iterators do not point to the same Sample.
+          bool operator !=(ConstIterator other) const;
+
+          //! Checks whether the Sample pointed to by the first iterator is to
+          //! the left of that pointed to by the second iterator.
+          bool operator <(ConstIterator other) const;
+
+          //! Checks whether the Sample pointed to by the first iterator is not
+          //! to the right of that pointed to by the second iterator.
+          bool operator <=(ConstIterator other) const;
+
+          //! Checks whether the Sample pointed to by the first iterator is to
+          //! the right of that pointed to by the second iterator.
+          bool operator >(ConstIterator other) const;
+
+          //! Checks whether the Sample pointed to by the first iterator is not
+          //! to the left of that pointed to by the second iterator.
+          bool operator >=(ConstIterator other) const;
+
+          //! Returns a reference to the Sample the iterator points to.
+          const Sample& operator *() const;
+
+          //! Returns a pointer to the Sample the iterator points to.
+          const Sample* operator ->() const;
+
+          //! Returns a reference to the Sample the iterator would point to
+          //! if it was advanced by the offset.
+          const Sample& operator [](std::ptrdiff_t offset) const;
```

Comment:
Remove this method from the iterator. If someone wants to index through an iterator they can use arithmetic.

Reply:
But operator[] is part of the RandomAccessIterator concept, isn't it? E.g.
auto v = std::vector<int>{1, 2};
auto x = v.begin()[1];
is perfectly valid.